File: src/shady/generator/generator_main.c
#include "generator.h"

#include "util.h"
#include "portability.h"

void preprocess(json_object* root);
void generate(Growy* g, json_object* root);

enum {
    ArgSelf = 0,
    ArgDstFile,
    ArgFirstInput,
};

int main(int argc, char** argv) {
    assert(argc > ArgFirstInput);
    int inputs_count = argc - ArgFirstInput;
    char* dst_file = argv[ArgDstFile];

    json_tokener* tokener = json_tokener_new_ex(32);
    enum json_tokener_error json_err;

    typedef struct {
        size_t size;
        char* contents;
        json_object* root;
    } JsonFile;

    LARRAY(JsonFile, json_files, inputs_count);
    for (size_t i = 0; i < inputs_count; i++) {
        String path = argv[ArgFirstInput + i];
        shd_read_file(path, &json_files[i].size, &json_files[i].contents);
        json_files[i].root = json_tokener_parse_ex(tokener, json_files[i].contents, json_files[i].size);
        json_err = json_tokener_get_error(tokener);
        if (json_err != json_tokener_success) {
            shd_error("Json tokener error while parsing %s:\n %s\n", path, json_tokener_error_desc(json_err));
        }

        shd_info_print("Correctly opened json file: %s\n", path);
    }
    Growy* g = shd_new_growy();

    json_object* src = json_object_new_object();

    for (size_t i = 0; i < inputs_count; i++) {
        json_apply_object(src, json_files[i].root);
    }

    preprocess(src);
    generate(g, src);

    size_t final_size = shd_growy_size(g);
    shd_growy_append_bytes(g, 1, (char[]) { 0 });
    char* generated = shd_growy_deconstruct(g);
    shd_debug_print("debug: %s\n", generated);
    if (!shd_write_file(dst_file, final_size, generated)) {
        shd_error_print("Failed to write file '%s'\n", dst_file);
        shd_error_die();
    }
    free(generated);
    for (size_t i = 0; i < inputs_count; i++) {
        free(json_files[i].contents);
        json_object_put(json_files[i].root);
    }
    json_object_put(src);
    json_tokener_free(tokener);
}

File: src/generator/generator_main.py
import json
import sys

from src.generator.generator import *
from src.generator.generator_common import preprocess
from src.generator.json_apply import json_apply_object

ArgSelf = 0,
ArgDstFile = 1
ArgFirstInput = 2

class JsonFile:
    contents = None
    root = None

def shd_read_file(path):
    f = open(path)
    return f.read()

def shd_write_file(path, contents):
    f = open(path, "w")
    f.write(contents)

def main(generate):
    argv = sys.argv
    argc = len(argv)

    dst_file = argv[ArgDstFile]

    json_files = []
    for path in argv[ArgFirstInput:]:
        json_file = JsonFile()
        json_file.contents = shd_read_file(path)
        json_file.root = json.loads(json_file.contents)
        json_files.append(json_file)
        print(f"Correctly opened json file: {path}")

    g = Growy()
    src = dict()

    for file in json_files:
        json_apply_object(src, file.root)

    preprocess(src)
    generate(g, src)

    shd_write_file(dst_file, g.g)